proposals/trusted_registry_crd.md

```diff
@@ -0,0 +1,128 @@
+# Introduce a custom resource to represent a (trusted) registry
+
+## TOC
+
+- [Abstract](#abstract)
+- [Problem statement](#problem-statement)
+- [Solution](#solutions)
+- [Use cases](#use-cases)
+  - [Starboard Security Operator supervising trusted registries](#starboard-security-operator-supervising-trusted-registries)
+  - [Repository webhooks integrated with Starboard](#repository-webhooks-integrated-with-starboard)
+
+## Abstract
+
+Common requirements for running secure Kubernetes workloads are:
+
+- Pull container images only from trusted registries and/or repositories. For example, one might allow pulling
+  images from private Harbor repository accessible at https://core.harbor.domain/library/nginx:1.16, but deny pull
+  requests to the core.harbor.domain/library/redis:5 or https://docker.io/library/nginx:1.16 repositories.
+- Statically scan container images for critical, high or medium vulnerabilities. In general, these rules are configurable.
+- Use a Kubernetes admission controller to deny workloads with non-compliant container images. For example, one may
+  allow `kubectl run nginx-pod --image=core.harbor.domain/library/nginx:1.16`, but deny
+  `kubectl run redis-pod --image core.harbor.domain/library/redis:5`.
+
+## Problem statement
+
+- Starboard CLI and Starboard Security Operator are able to scan workloads already deployed in a cluster. Hence, an 
+  admission controller cannot use vulnerability reports for compliance checks.
+- The `vulnerabilities.aquasecurity.github.io` resources are scoped to the same namespace as the underlying workload.
+  However, for admission controllers we might want to introduce cluster-scoped resource to represent vulnerability
+  report associated with a given container image digest, for example `registryvulnerabilities.aquasecurity.github.io`.
+- Container images are not first-class citizens in Kubernetes-native world. There’s no built-in resource that represents
+  a container image. Also, there’s no built-in representation of a (trusted) registry.
+- An admission controller has certain performance requirements. Usually we cannot trigger a scanner when the admission
+  controller is invoked for a given Kubernetes API request. Most likely it will take way too long to perform such
+  ad hoc scans.
+
+## Solutions
+
+### Define registries.aquasecurity.github.io resource
+
+First off, introduce a custom resource to represent a (trusted) container registry. For example, we can represent
+a private Harbor registry with the following Kubernetes object:
+
+```yaml
+apiVersion: aquasecurity.github.io/v1alpha1
+kind: Registry
+metadata:
+  name: my-harbor-registry
+spec:
+  name: My Harbor Registry
+  description: |
+    That's the only registry authorized to run production workloads.
+  server: core.harbor.domain
+  imagePullSecrets:
+  - name: my-harbor-registry-credentials
+  insecure: false
+  trustedRepositories:
+  - name: library/nginx:1.16
```

Comment:
Do you think we need to restrict down to the level of approved / not-approved tags? I was imagining more at the repo level. Or even just to the registry level. 

Registries - only allow images from registries that have a CRD instance in place 
Repositories - do they have to have a trustedRepositories entry? Or maybe the default is to allow everything from all repos in a trusted registry? (I am not totally convinced we need the granularity of repository-level control for the first version...) 

Edit: on further reading I think you are suggesting using this as the list of repos that the Starboard operator knows to scan. I wonder if this is conflating two different things. We could have a CRD for trusted registries, and a Starboard config map for a list of repos to be scanned? I am not sure if this is better, just raising it as an option. 

Having a CRD for trusted registries would allow for a super-simple admission controller that denies images that don't come from a trusted registry.

Reply:
- The first use case for such CRD that came to my mind was an admission controller that required a container image to be already scanned. For example, the admission controller could reject a deployment if:
  - the container image hasn't been scanned
  - the container image has already been scanned but has a critical vulnerability
- In other words, my assumption was that we do reject deployments by default, unless they're scanned
- The other use case for the registries CRD is to specify the desired state of which container images we want to be scanned. Whereas the actual state is represented by the corresponding vulnerability reports. So the desired state could be reconciled by the Starboard Security Operator if the integration via webhooks was not possible

Reply:
And I agree, we should think more about the granularity of such allow list, whether it's registry, repository, or tag/digest level.